chips/sifive/src/plic.rs

```diff
@@ -10,36 +10,59 @@ use kernel::utilities::registers::LocalRegisterCopy;
 use kernel::utilities::registers::{register_bitfields, ReadOnly, ReadWrite};
 use kernel::utilities::StaticRef;
 
+///    The generic SiFive PLIC specification:
+///    https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc
+///    is defining maximum of 1023 interrupt sources
+
+const MAX_INTERRUPTS: usize = 1023;
+/// maximum number of bit-coded registers, 1 bit per interrupt
+const MAX_BIT_REGS: usize = MAX_INTERRUPTS.div_ceil(32);
+
+/// PLIC registers for *machine mode* context only at this time.
+/// The spec defines extra sets of registers for additional contexts,
+/// that is supervisor, user and other modes, but these aren't supported
+/// by the current code.
+
 #[repr(C)]
 pub struct PlicRegisters {
     /// Interrupt Priority Register
     _reserved0: u32,
-    priority: [ReadWrite<u32, priority::Register>; 51],
-    _reserved1: [u8; 3888],
+    priority: [ReadWrite<u32, priority::Register>; MAX_INTERRUPTS],
+    _reserved1: [u8; 0x1000 - (MAX_INTERRUPTS + 1) * 4],
```

Comment:
This makes me nervous. Generally, we try to trend towards static, compile-time correctness of hardware accesses; it seems the lack of cost generics forces(?) us to do differently here?

While we are generally strongly adverse to `cfg` flags, this _may_ be a case where they are appropriate, as the `cfg` is fully unambiguous and necessary(?) for correctly representing a hardware target.

I would like someone more in the know of the RISC-V ecosystem and design paradigms (@lschuermann ?) to weigh in before advocating for a `cfg` here, however.

In the meantime, the related question is what happens if there is a read or write to a non-implemented PLIC register? Hang? Exception? Undefined?

Reply:
> In the meantime, the related question is what happens if there is a read or write to a non-implemented PLIC register? Hang? Exception? Undefined?
That's a legit question. I'd expect nothing to happen, as there is no appropriate exception which would properly describe this (it is not a PMP access error, it is also not a bus error, which would fire in case we have a BEU in the configuration). But unfortunately I can't find any definition in the specs to claim that it is well defined.

Reply:
I agree this is a time where a `cfg` is appropriate. Any chips that instantiate this PLIC should know exactly how many interrupts are defined and the registers object should match what exists in hardware.

Reply:
I'm not sure how you would use a `cfg` flag well here. My understanding is that any value between 0-1023 is possible for different implementations of this IP and we're certainly not going to have 1023 different `cfg` flags. I don't believe it's possible to pass values in `cfg` and, even if it were, it doesn't solve the problem of needing to compute the amount reserved registers after the priority registers from a compile-time value.

In any case, I also believe this is all moot using @jrvanwhy's new Tock registers proposal and we're better with some imperfection in the internals here now rather than adding a bunch of complexity for later with relatively limited benefit now.

Reply:
We don't need to support all 1023, just all that hardware actually implements (and I guess that we care to support).

Reply:
I think the problem, though, is that the sifive chip is not _actually_ a specific discreet set of chips, but rather hardware IP that people then customize for specific chips (the e310x is an _example_ of such a chip upstream). So, in cfg strategy, anyone who uses the sifive IP either upstreams their chip (which we basically wouldn't want since it would eventually mean 1023 different cfg features) _or_ they fork the sifive chip crate to customize the priority registers size.

Conversely, if either the priority registers uses `MAX_INTERRUPTS` _or_ has a way of using something like const generics to specialize the number of priority registers, a downstream chip needs to do neither.

Reply:
FWIW, I am working with 129 interrupts config at the moment, and we have few more in the plans. hat's right, it isn't a commonly available platform, but it would be nice to have out-of-the box support.

Reply:
We actually get no stronger compile-time guarantees atm than we would from this PR current changes because of how `priority` is accessed and abstracted. Iterating over `priority` as `master` does is constrained by the `total_ints` value in the PR, and more importantly, individual access to priorities would be checked and might panic because the offset is not constant inline, so there would be a dynamic against the length that can panic.

I think my suggestion below is general while getting as much compile-time safety as the cfg version. I still think both are too complex or niche to be worth it.

_However_ I won't stand in the way of this kind of cfg if that's what's preferred by others.

Reply:
I guess we *could* do something like this:
```rust
#[cfg(not(feature="custom_plic))]
const MAX_INTERRUPTS: usize = 51;
#[cfg(feature="custom_plic)]
include!("path/to/plic/config/rs/file/defining/MAX_INTERRUPTS");
```

or

```rust
#[cfg(not(feature="custom_plic))]
const MAX_INTERRUPTS: usize = 51;
#[cfg(feature="custom_plic)]
include_str!(concat!("const MAX_INTERRUPTS: usize = "), env!("CUSTOM_PLIC_INT_NUM"));
```
and have a `build.rs` defining the env variable.
But both options seem to be equally horrible to me :D

Reply:
I don't think in this case we need any of the extra complexities here.

In practice constant propagation and inlining will reduce this to the same as what we already have, and the getters enforce an interface that won't allow naive users to accidentally access non existent hardware.

Normally I'm not one to advocate for 'the optimizer will handle it', but all the workarounds just add a lot of complexity for no real benefit here as far as I can see. If/when const generics land, we can rewrite this to explicitly do what the optimizer is implicitly doing.

Maybe add a "todo: replace with const generic when stable", but otherwise I think this is good as is